### ISSUE TITLE
Is it possible to use Dlisio to read dlis files from other contractors (Halliburton, Baker)

### ISSUE THREAD

**Issue by veritolilo** [equinor/dlisio#174]
Hi,  I am a new user of the dlisio library, which I found extremely useful to read and extract information from.  However, I am working with dlis files from different contractors beside Schlumberger and I have not been able to read these other dlis files.  I was wondering if the library works only for Schlumberger dlis or I am doing something wrong.  Thank in advance for your comments and clarifications.

This is the simple code for reading a file with their corresponding error:

with dl.load('Well_16_1_9/WL_RAW_AC-DEN-GR-NEU-REMP_MWD_1.DLIS') as files:
    for f in files:
        pass

Traceback (most recent call last):

  File "<ipython-input-34-8f6deef34285>", line 1, in <module>
    with dl.load('Well_16_1_9/WL_RAW_AC-DEN-GR-NEU-REMP_MWD_1.DLIS') as files:

  File "C:\Users\veront\AppData\Local\Continuum\anaconda3\lib\site-packages\dlisio\__init__.py", line 596, in load
    tells, residuals, explicits = core.findoffsets(mmap, vrlpos)

RuntimeError: record-length in record 3 corrupted

**Comment by jokva** [comment]
Hi!

dlisio not Schlumberger specific at all, so other contractors should work - in fact, we have successfully read some reference Baker files.

Now, to your issue, something goes wrong of course. This is either a spec violation in your file, or a bug in dlisio. One thing is clear, you're not doing it wrong.

Now onto the tricky part: actually figuring out what's going wrong here. The by *far* easiest way is to give us a copy of the file, so us developers can manually inspect and trace the issue. If that works for you, that would be the easiest solution for all parties.

While I can promise you we're not at all interested in the contents, we _do_ understand if that's simply not possible. If so, we'll have to look really hard at the spec and implementation, to figure out what's causing it and be able to reproduce it.

**Comment by veritolilo** [comment]
Hei, 

Thank you for your response. I will confirm if I can send the file to you. So it would be much faster to detect the problem and got a feedback.

**Comment by veritolilo** [comment]
Hi, 
Here is the dlis file. This data is public domain. 
[WL_RAW_AC-DEN-GR-NEU-REMP_MWD_1.zip](https://github.com/equinor/dlisio/files/3716777/WL_RAW_AC-DEN-GR-NEU-REMP_MWD_1.zip)

**Comment by jokva** [comment]
Excellent, thanks!

Where is this file from, and what is the source? Can we find more licence information somewhere?

If possible, we'd like to keep a copy in the repository.

**Comment by veritolilo** [comment]
Your welcome, 

The file is from Halliburton MWD in the Ivar Aasen field (north sea), and got it from Diskos database (NPD).

**Comment by veritolilo** [comment]
Hei,

Were you able to read the Halliburton dlis file?

**Comment by ErlendHaa** [comment]
Hi

After investigating the file, we found that the issue is related to #179. In short, you file contains "trash"-bytes in-between the actual data. As far as we can tell at this point, these bytes seem to be what the standard [1] refers to as tapemarks. 

Unfortunately, dlisio are not able to read files with tapemarks at the moment

It might seem like this is a fairly common issue that we need to address. However, the standard is _very_ wage on where they might occur and how they look like. Hence, before we can offer official support for files with tapemarks, we need to do some further investigation.

**Comment by veritolilo** [comment]
Hei,
I understand the problem. Ok so at the moment is not possible to work with
those files by using dlisio.  =(

thank you very much for your response.

**Comment by ejschoen** [comment]
I have a somewhat critical need to deal with Halliburton Insite DLIS files that have the problem described in this issue.  To see if I could make progress, I tweaked the dlisio.cpp and io.cpp code so that when a logical record appears to have a length < 20, the code skips 12 bytes and re-reads the VRL.  (And then if the length is < 20 or the version is not 1, the code gives up as before.)  This appears to work for these files.  At least I am able to get past the point in __init.py__'s `load` function where the C++ `findoffsets` is called.   

I get into the call to `core.findfdata` in `load`, and it appears to me that the C++ code constructs a good dictionary from DLIS frame obnames to record numbers.  But somewhere between the C++ and Python worlds, Pybind11 chokes on some Unicode conversion:

```
  File "***DLIS2Text.py", line 249, in generateText
    with dlisio.load(filename) as files:
  File "/***/.local/lib/python3.7/site-packages/dlisio/__init__.py", line 783, in load
    part['implicits'], part['tells'], part['residuals']):
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 13: invalid start byte
```

If I look at what `findfdata` is trying to return, it looks like this:
```
(lldb) print xs
(std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int> > >) $31 = size=22815 {
  [0] = (first = "T.FRAME-I.52-O.2-C.0", second = 12)
  [1] = (first = "T.FRAME-I.52-O.2-C.0", second = 13)
  [2] = (first = "T.FRAME-I.52-O.2-C.0", second = 14)
  [3] = (first = "T.FRAME-I.52-O.2-C.0", second = 15)
...
 [255] = (first = "T.FRAME-I.52-O.2-C.0", second = 267)
  ...
}
```
Assuming that the Unicode in question comes from the first field of `std::pair<std::string,int>`, I am not seeing an issue (and I've used parray to print each element of the 18598-entry `xs` vector.   

Any thoughts on Unicode issues with Halliburton Insite files?  I've tried 3 different ones, and they all have the same problem.  I have no problem with a Weatherford DLIS file (but that file doesn't have tape marks, either.)

Unfortunately, the files I'm processing are customer proprietary, so I probably cannot supply them.

**Comment by ErlendHaa** [comment]
Hi

The next release of dlisio will include support for files with tapemarks. However I cannot give you an exact time frame for when that will be.

If, in the mean time you want to get your approach up and running I can offer you some pointers. 

> when a logical record appears to have a length < 20, the code skips 12 bytes and re-reads the VRL

The 12 bytes you are skipping is a tapemark, they can occur before, after and in between Visible Records.  However their presence has nothing to do with the record lengths [1].

Rather than checking VRL < 20, you want to check the *format field* of the VR. For any DLISv1 file this should be `0xFF 0x01` [2]. If the parsed format field is something else than that, you can skip 12 bytes and try again. 

However, this is **not** guaranteed to always work. But if you have a a small set of similar files this approach might work, maybe with some additional changes aswell.

[1] The restriction that VRL must be at least 20 bytes is governed by the dlis spec but we have seen files that violates this. However that's a separate issue to what you are trying to solve here. 

[2] You might notice that the unicode error throws on the byte `0xFF`. I expect that this is from the format field. I.e. the approach taken in findoffsets, has yielded wrong offsets, restulting findfdata to to create the fingerprint from the wrong set of bytes.

**Comment by ejschoen** [comment]
Thanks!  I'll look into findoffsets and see if there is something obvious.  In the interim, I had added an option to dlisio.load to skip loading frame data, which at least allowed me to get enough static data to determine what was in the files.  But without frame data, I can't determine index_min and index_max for frames.

**Comment by ErlendHaa** [comment]
`index_min` and `index_max` is actually recorded in the static data, more precisely on the `Frame`-objects.  https://dlisio.readthedocs.io/en/stable/dlisio.html#frame

**Comment by ErlendHaa** [comment]
About the behaviour you observe regarding the unicode error

Creating the `std::string` for the vector of pair in `findfdata` works fine because `std::string` does not require the bytes to be valid ascii characters. However, a python string _does_. Hence, pybind fails when it tries to cast from `std::string` to python strings.

**Comment by ejschoen** [comment]
Right.  What I've got works great apparently until residuals[i]==0 in findfdata. At that point, tells is off by 12 or 16 bytes.  I get past one of these, because dlis_obname sees a 0-byte id length and doesn't change the stack-allocated name array in findfdata. (dlis_obname in types.cpp could probably throw on idlen==0... are 0-length idents legal?).  Ultimately, I run into an IFLR at the top of a logical record and tells is 12 bytes too small, and I suck in the ff byte of the VRL as part of the string.  So, as you suggested, some logic I added in findoffsets is probably still off in (not) readjusting tells for the tape marks.
